<a href="https://colab.research.google.com/github/mdf04414/mdf04414/blob/main/CO_Assignment_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Unsigned Adder
Our objective is to understand and build an Adder that is able to add unsigned bits. 

## Question - 1
Write Python code that does the following. 

- Create the following functions, that will return the appropriate response as mentioned below for a given input bit.
    - function: **and_gate()**, Input: **input_value_a**, **input_value_b** where input_value_a and input_value_b should be unsigned four bit binary value but they may not always have 4 bits in which case we prefix the required number of 0's to make it of length 4, Ex: 01 is considered as 0001 This function should return the appropraite value as an AND GATE operation on the following inputs would for the given two input bits
        - Ex: input_value_a = 1110, input_value_b = 1010 then input_value_a AND input_value_b would result in output_value = 1010
    - function: **or_gate()**, Input: **input_value_a**, **input_value_b** where input_value_a and input_value_b should be unsigned four bit binary value but they may not always have 4 bits in which case we prefix the required number of 0's to make it of length 4, Ex: 01 is considered as 0001 This function should return the appropraite value as an OR GATE operation on the following inputs would for the given two input bits
        - Ex: input_value_a = 1110, input_value_b = 1010 then input_value_a OR input_value_b would result in output_value = 1110
    - function: **xor_gate()**, Input: **input_value_a**, **input_value_b** where input_value_a and input_value_b should be unsigned four bit binary value but they may not always have 4 bits in which case we prefix the required number of 0's to make it of length 4, Ex: 01 is considered as 0001 This function should return the appropraite value as an XOR GATE operation on the following inputs would for the given two input bits
        - Ex: input_value_a = 1110, input_value_b = 1010 then input_value_a XNOR input_value_b would result in output_value = 0100
    - function: **not_gate()**, Input: **input_value** where input_value should be unsigned N bit binary value. This function should return the appropraite value as an NOT GATE operation on the following inputs would for the given input bits
        - Ex: input_value = 110, then result in output_value = 001

**Conditions to follow:**
- The output should be a string format similar to test cases mentioned below
- **IMPORTANT: You cannot use any Python [mathematical operators](https://www.w3schools.com/python/gloss_python_arithmetic_operators.asp) in your solution.**
- Inbuilt functions which require no import and Datatype methods can be used to solve this problem but No arbitray functions or library can be used.
- No other function name can be defined/declared than the ones asked to be declared in any of the questions presented in this notebook.



---

TEST CASES:

```
Input A: 01 AND Input B: 1111 is 0001

Input A: 01 OR Input B: 1111 is 1111

Input A: 01 XOR Input B: 1111 is 1110

Input A: 01 NOT is 10, Input B: 1111 NOT is 0000

Input A: 1010 AND Input B: 101 is 0000

Input A: 1010 OR Input B: 101 is 1111

Input A: 1010 XOR Input B: 101 is 1111

Input A: 1010 NOT is 0101, Input B: 101 NOT is 010
```
---

Hint: You can use built in methods and data type/string methods to create the solution to solve this problem.
You can come up with a solution that does not use this hint but it should satisfy all the TEST CASES.

### Code

In [ ]:
# DECLARE the function and_gate()
def and_gate(input_value_a, input_value_b):
    # We will add zeros to the MSB if the length is less than 4
    input_value_a = input_value_a.zfill(4)
    input_value_b = input_value_b.zfill(4)
    # Initialize the output_value_list
    output_value_list = []
    # Iterate over both input, bit by bit
    for bit_a, bit_b in zip(input_value_a, input_value_b):
        # AND GATE output is 1 only when both inputs are 1 and 0 otherwise
        output_value_list.append('1' if (bit_a == '1') and (bit_b == '1') else '0')
    # Return the output value as string        
    return "".join(output_value_list)

# DECLARE the function or_gate()
def or_gate(input_value_a, input_value_b):
    # We will add zeros to the MSB if the length is less than 4
    input_value_a = input_value_a.zfill(4)
    input_value_b = input_value_b.zfill(4)
    # Initialize the output_value_list
    output_value_list = []
    # Iterate over both input, bit by bit
    for bit_a, bit_b in zip(input_value_a, input_value_b):
        # OR GATE output is 1 only when either inputs are 1 and 0 otherwise
        output_value_list.append('1' if (bit_a == '1') or (bit_b == '1') else '0')
    # Return the output value as string        
    return "".join(output_value_list)

# DECLARE the function xor_gate()
def xor_gate(input_value_a, input_value_b):
    # We will add zeros to the MSB if the length is less than 4
    input_value_a = input_value_a.zfill(4)
    input_value_b = input_value_b.zfill(4)
    # Initialize the output_value_list
    output_value_list = []
    # Iterate over both input, bit by bit
    for bit_a, bit_b in zip(input_value_a, input_value_b):
        # XOR GATE output is 1 only when both inputs are not equal to each other and 0 otherwise
        output_value_list.append('1' if (bit_a != bit_b) else '0')
    # Return the output value as string        
    return "".join(output_value_list)

# DECLARE the function not_gate()
def not_gate(input_value):
    # Switch the places of 1 and 0 in the input value string
    input_value = input_value.replace('1', 'X').replace('0', '1').replace('X', '0')
    # Return the output value as string        
    return input_value

### TEST CASES ###

# Initialize the input values
input_value_tuple_list = [('01', '1111'), ('1010', '101')]
for input_value_a, input_value_b in input_value_tuple_list:
        # CALL the and_gate with required parameters
    and_gate_output_value = and_gate(input_value_a=input_value_a, input_value_b=input_value_b)
    # CALL the or_gate with required parameters
    or_gate_output_value = or_gate(input_value_a=input_value_a, input_value_b=input_value_b)
    # CALL the xor_gate with required parameters
    xor_gate_output_value = xor_gate(input_value_a=input_value_a, input_value_b=input_value_b)
    # CALL the not_gate with required parameters
    not_gate_a_output_value = not_gate(input_value=input_value_a)
    not_gate_b_output_value = not_gate(input_value=input_value_b)
    # Print the output
    print("Input A: {0} AND Input B: {1} is {2}\n".format(input_value_a, input_value_b, and_gate_output_value))
    print("Input A: {0} OR Input B: {1} is {2}\n".format(input_value_a, input_value_b, or_gate_output_value))
    print("Input A: {0} XOR Input B: {1} is {2}\n".format(input_value_a, input_value_b, xor_gate_output_value))
    print("Input A: {0} NOT is {2}, Input B: {1} NOT is {3}\n".format(input_value_a, input_value_b, not_gate_a_output_value, not_gate_b_output_value))

Input A: 01 AND Input B: 1111 is 0001

Input A: 01 OR Input B: 1111 is 1111

Input A: 01 XOR Input B: 1111 is 1110

Input A: 01 NOT is 10, Input B: 1111 NOT is 0000

Input A: 1010 AND Input B: 101 is 0000

Input A: 1010 OR Input B: 101 is 1111

Input A: 1010 XOR Input B: 101 is 1111

Input A: 1010 NOT is 0101, Input B: 101 NOT is 010



## Question - 2
Write Python code that does the following. 

- Create a function: **four_bit_unsigned_adder()**, Input: **input_value_a**, **input_value_b** where input_value_a and input_value_b should be unsigned four bit binary value but they may not always have 4 bits in which case we prefix the required number of 0's to make it of length 4, Ex: 01 is considered as 0001 This function should ADD the four bits and return the appropraite value. Note: The examples below show the decimal equivalent of the binary input inside paranthesis and their sum will be equal to the output_value binary which also when in decimal will be same the decimal equivalent of the input value addition.
    - Ex: input_value_a = 1000 (8), input_value_b = 1010 (10) then input_value_a ADD input_value_b would result in output_value = 10010 (18)
    - Ex: input_value_a = 1000 (8), input_value_b = 010 (2) then input_value_a ADD input_value_b would result in output_value = 01010 (10)

**Conditions to follow:**
- The output should be a string format similar to test cases mentioned below
- **IMPORTANT: You cannot use any Python [mathematical operators](https://www.w3schools.com/python/gloss_python_arithmetic_operators.asp) in your solution.**
- Inbuilt functions which require no import and Datatype methods can be used to solve this problem but No arbitray functions or library can be used.
- No other function name can be defined/declared than the ones asked to be declared in any of the questions presented in this notebook. 


---

TEST CASES:

```
Input A: 1000 ADDED Input B: 01 is 1001

Input A: 1010 ADDED Input B: 0101 is 1111

Input A: 1110 ADDED Input B: 0111 is 10101

Input A: 010 ADDED Input B: 111 is 1001
```
---

Hint: You can use built in methods and data type/string methods to create the solution to solve this problem. Consider using logic gates from question-1 and building a parallel adder. 


### Code

In [ ]:
# DECLARE the function four_bit_unsigned_adder()
def four_bit_unsigned_adder(input_value_a, input_value_b):
  # initializing a string to display the result
    res = ""
    # initializing a carry variable
    carry = 0

    # reversing both input a and b because you want to add them right to left instead of left to right
    input_value_a, input_value_b = input_value_a[::-1], input_value_b[::-1]
    
    #iterating over the maximum length of either input variable a or b, depends which one is longer
    #this will make it so that so that it will always end up in max bits
    for x in range(max(len(input_value_a), len(input_value_b))):

      #if x is less than the input value lenght, it will make basically take the string of the number and turns it into an integer, else, it will add 0
      digitA = ord(input_value_a[x]) - ord("0") if x < len(input_value_a) else 0
      digitB = ord(input_value_b[x]) - ord("0") if x < len(input_value_b) else 0

      #this total variable essentially takes a + b + carry
      total = digitA + digitB + carry
      char = str(total % 2) # 0 + 1 + 0 = 1 1%2 = 1 which is the sum.  1+1+0 = 2 % 2 = 0 which is the sum.  Using str method bc cant add with integers
      res = char + res #basically appending the string of each digit to res
      carry = total // 2 #1+1 = 2. 2//2 = 1 (which is the carry)
    if carry:
      res = "1" + res # basically adding 1 to the result if there is a carry
    return res
 

### TEST CASES ###

# Initialize the input values
input_value_tuple_list = [('1000', '01'), ('1010', '0101'), ('1110', '0111'), ('010', '111')]
for input_value_a, input_value_b in input_value_tuple_list:
    # CALL the four_bit_unsigned_adder with required parameters
    output_value = four_bit_unsigned_adder(input_value_a=input_value_a, input_value_b=input_value_b)
    # Print the output
    print("Input A: {0} ADDED Input B: {1} is {2}\n".format(input_value_a, input_value_b, output_value))

Input A: 1000 ADDED Input B: 01 is 1001

Input A: 1010 ADDED Input B: 0101 is 1111

Input A: 1110 ADDED Input B: 0111 is 10101

Input A: 010 ADDED Input B: 111 is 1001



# Subtractor
Our objective is to understand and build Subtractor and Comparator that is able to add unsigned bits.

## Question - 3
Write Python code that does the following. 

- Create a function: **four_bit_unsigned_subtractor()**, Input: **input_value_a**, **input_value_b** where input_value_a and input_value_b should be unsigned four bit binary value but they may not always have 4 bits in which case we prefix the required number of 0's to make it of length 4, Ex: 01 is considered as 0001 This function should SUBTRACT the four bits and return the appropraite value, Here the output can be both a positive and negetive binary value. Note: The examples below show the decimal equivalent of the binary input inside paranthesis and their sum will be equal to the output_value binary which also when in decimal will be same the decimal equivalent of the input value addition. **Also not we will show the '+' or '-' symbol before the binary value to indicate if it a positive or negative value**

    - Ex: input_value_a = 1000 (8), input_value_b = 1010 (10) then input_value_a SUBTRACT input_value_b would result in output_value = -0010 (-2)
    - Ex: input_value_a = 1000 (8), input_value_b = 010 (2) then input_value_a ADD input_value_b would result in output_value = +0110 (6)

**Conditions to follow:**
- The output should be a string format similar to test cases mentioned below
- **IMPORTANT: You cannot use any Python [mathematical operators](https://www.w3schools.com/python/gloss_python_arithmetic_operators.asp) in your solution.**
- Inbuilt functions which require no import and Datatype methods can be used to solve this problem but No arbitray functions or library can be used.
- No other function name can be defined/declared than the ones asked to be declared in any of the questions presented in this notebook. 
- The code you write should pass all the test cases given. 


---

TEST CASES:

```
Input A: 111 SUBTRACTED Input B: 010 is +0101

Input A: 1010 SUBTRACTED Input B: 0101 is +0101

Input A: 1110 SUBTRACTED Input B: 0111 is +0111

Input A: 010 SUBTRACTED Input B: 111 is -101
```
---

Hint: You can use built in methods and data type/string methods to create the solution to solve this problem. Consider using logic gates from question-1 and building a parallel subtractor based on the logic of the parallel adder. 

### Code

In [ ]:
# DECLARE the function four_bit_unsigned_subtractor()
def four_bit_unsigned_subtractor(input_value_a, input_value_b):
    # We will use a 4-bit parallel adder logic
    input_value_a = input_value_a.zfill(4)
    input_value_b = input_value_b.zfill(4)
    # We will create 1's compliment of the input_value_b using by using XOR gate on each bit with 1
    input_value_b = "".join([xor_gate(bit_b, '1')[-1] for bit_b in input_value_b])
    # Initialize the output_value_list
    output_value_list = []
    # Initialize the carry from each addition which is initally 1 which should make the input_value_b into 2's compliment
    carry = '1'
    # Iterate over each bit and add them, We need to start operating from LSB to MSB
    for bit_a, bit_b in zip(input_value_a[::-1], input_value_b[::-1]):
        # SUM = A XOR B XOR Ci
        sum = xor_gate(xor_gate(bit_a, bit_b)[-1], carry)[-1]
        # Insert Sum at MSB
        output_value_list.insert(0, sum)
        # CARRY = A.B + B.C + C.A
        a_and_b, b_and_c, c_and_a = and_gate(bit_a, bit_b)[-1], and_gate(bit_b, carry)[-1], and_gate(carry, bit_a)[-1]
        carry = or_gate(or_gate(a_and_b, b_and_c)[-1], c_and_a)[-1]
    if carry == '0':
        output_value_list = [xor_gate(output_bit, '1')[-1] for output_bit in output_value_list]
        output_value_list = list(four_bit_unsigned_adder(input_value_a="".join(output_value_list), input_value_b='1')[-4:])
        output_value_list.insert(0, '-')
    else:
        output_value_list.insert(0, '+')
    # Return the output value as string
    return "".join(output_value_list)

### TEST CASES ###

# Initialize the input values
input_value_tuple_list = [('111', '010'), ('1010', '0101'), ('1110', '0111'), ('010', '111')]
for input_value_a, input_value_b in input_value_tuple_list:
    # CALL the four_bit_unsigned_subtractor with required parameters
    output_value = four_bit_unsigned_subtractor(input_value_a=input_value_a, input_value_b=input_value_b)
    # Print the output
    print("Input A: {0} SUBTRACTED Input B: {1} is {2}\n".format(input_value_a, input_value_b, output_value))

Input A: 111 SUBTRACTED Input B: 010 is +0101

Input A: 1010 SUBTRACTED Input B: 0101 is +0101

Input A: 1110 SUBTRACTED Input B: 0111 is +0111

Input A: 010 SUBTRACTED Input B: 111 is -0101

